### PR TITLE
Use height of latest network upgrade in -mineraddress validation.

### PR DESCRIPTION
**Pull request by therealyingtong**
Closes https://github.com/zcash/zcash/issues/5778

### CODE REVIEW

File: src/consensus/params.h [hunk 1]
```diff
@@ -237,6 +237,11 @@ struct Params {
      */
     bool NetworkUpgradeActive(int nHeight, Consensus::UpgradeIndex idx) const;
 
+    /**
+     * Returns the activation height of the latest confirmed upgrade.
+     */
+    int HeightOfLatestConfirmedUpgrade() const;
```
Comment (daira) [inline src/consensus/params.h:243]:
```suggestion
     * Returns the activation height of the latest settled upgrade.
     */
    int HeightOfLatestSettledUpgrade() const;
```

File: src/consensus/params.cpp [hunk 2]
```diff
@@ -25,6 +25,15 @@ namespace Consensus {
         return NetworkUpgradeState(nHeight, *this, idx) == UPGRADE_ACTIVE;
     }
 
+    int Params::HeightOfLatestConfirmedUpgrade() const {
```
Comment (daira) [inline src/consensus/params.cpp:28]:
```suggestion
    int Params::HeightOfLatestSettledUpgrade() const {
```

File: src/init.cpp [hunk 3]
```diff
@@ -1099,9 +1099,11 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 #ifdef ENABLE_MINING
     if (mapArgs.count("-mineraddress")) {
         auto addr = keyIO.DecodePaymentAddress(mapArgs["-mineraddress"]);
-        if (!(addr.has_value() && std::visit(ExtractMinerAddress(chainparams.GetConsensus(), 0), addr.value()).has_value())) {
+        auto consensus = chainparams.GetConsensus();
+        int height = consensus.HeightOfLatestConfirmedUpgrade();
```
Comment (daira) [inline src/init.cpp:1103]:
```suggestion
        int height = consensus.HeightOfLatestSettledUpgrade();
```

File: src/init.cpp [hunk 4]
```diff
@@ -1099,9 +1099,11 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
+        if (!(addr.has_value() && std::visit(ExtractMinerAddress(consensus, height), addr.value()).has_value())) {
             return InitError(strprintf(
-                _("Invalid address for -mineraddress=<addr>: '%s' (must be a Sapling or transparent P2PKH address)"),
+                _("Invalid address for -mineraddress=<addr>: '%s' (must be a Sapling or transparent P2PKH address, or a Unified Address containing a valid receiver for the most recent network upgrade.)"),
```
Comment (daira) [inline src/init.cpp:1106]:
```suggestion
                _("Invalid address for -mineraddress=<addr>: '%s' (must be a Sapling or transparent P2PKH address, or a Unified Address containing a valid receiver for the most recent settled network upgrade.)"),
```

**Review by daira — COMMENTED** [review]
Added suggestions, also this is failing tests.

**Review by str4d — APPROVED** [review]
utACK

**Comment by str4d** [comment]
One of the failing builders failed in `wallet_unified_change.py`:
```
wallet_unified_change.py:
Initializing test directory /tmp/testal0eeqgq/176
waiting for async operation opid-baec62ed-950f-4a5c-9a9c-4df50f4d0403
waiting for async operation opid-9de19820-fa53-407a-b368-0318534f3d6d
Assertion failed: (left == right)
  left: <{'pools': {'sapling': {'valueZat': 1000000000}, 'orchard': {'valueZat': 1000000000}}, 'minimum_confirmations': 1}>
 right: <{'pools': {'orchard': {'valueZat': 1000000000}}, 'minimum_confirmations': 1}>
Stopping nodes
Cleaning up
Failed

stderr:
  File "/workspace/source/qa/rpc-tests/test_framework/test_framework.py", line 146, in main
    self.run_test()
  File "/workspace/source/qa/rpc-tests/wallet_unified_change.py", line 57, in run_test
    self.nodes[0].z_getbalanceforaccount(account0))
  File "/workspace/source/qa/rpc-tests/test_framework/util.py", line 566, in assert_equal
    raise AssertionError("(left == right)%s\n  left: <%s>\n right: <%s>" % (message, str(expected), str(actual)))
```

This looks like the bug that @nuttycom fixes in #5834, as it's during the test setup where we create two transactions to fund the test (one Sapling, one Orchard), and only one of them gets mined.

**Pull request merged**